Add forward AD for linalg.det and simplify its backward

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #80217
* #80074
* #80073
* #79838
* #79742
* __->__ #79487

This PR is in preparation for implementing `logdet` and `slogdet` as
structured kernels + implementing them with more efficient derivatives

We implement forward AD for det. We also simplify the implementation of
the backward, and leave a note on how to implement it properly for
singular matrices. We leave thad for future work.

Note (by looking at the OpInfo) that the current implementation passes
the same tests as the one before. We skip the forward-over-backward in
the singular case, as that one was not working in the gradgrad case
either.

lezcano committed:
Add forward AD for linalg.det and simplify its backward

This PR is in preparation for implementing `logdet` and `slogdet` as
structured kernels + implementing them with more efficient derivatives

We implement forward AD for det. We also simplify the implementation of
the backward, and leave a note on how to implement it properly for
singular matrices. We leave thad for future work.

Note (by looking at the OpInfo) that the current implementation passes
the same tests as the one before. We skip the forward-over-backward in
the singular case, as that one was not working in the gradgrad case
either.

[ghstack-poisoned]

lezcano committed:
Update on "Add forward AD for linalg.det and simplify its backward"


This PR is in preparation for implementing `logdet` and `slogdet` as
structured kernels + implementing them with more efficient derivatives

We implement forward AD for det. We also simplify the implementation of
the backward, and leave a note on how to implement it properly for
singular matrices. We leave thad for future work.

Note (by looking at the OpInfo) that the current implementation passes
the same tests as the one before. We skip the forward-over-backward in
the singular case, as that one was not working in the gradgrad case
either.

[ghstack-poisoned]

lezcano committed:
Update on "Add forward AD for linalg.det and simplify its backward"


This PR is in preparation for implementing `logdet` and `slogdet` as
structured kernels + implementing them with more efficient derivatives

We implement forward AD for det. We also simplify the implementation of
the backward, and leave a note on how to implement it properly for
singular matrices. We leave thad for future work.

Note (by looking at the OpInfo) that the current implementation passes
the same tests as the one before. We skip the forward-over-backward in
the singular case, as that one was not working in the gradgrad case
either.

[ghstack-poisoned]

lezcano committed:
Update on "Add forward AD for linalg.det and simplify its backward"


This PR is in preparation for implementing `logdet` and `slogdet` as
structured kernels + implementing them with more efficient derivatives

We implement forward AD for det. We also simplify the implementation of
the backward, and leave a note on how to implement it properly for
singular matrices. We leave thad for future work.

Note (by looking at the OpInfo) that the current implementation passes
the same tests as the one before. We skip the forward-over-backward in
the singular case, as that one was not working in the gradgrad case
either.

[ghstack-poisoned]

lezcano committed:
Update on "Add forward AD for linalg.det and simplify its backward"


This PR is in preparation for implementing `logdet` and `slogdet` as
structured kernels + implementing them with more efficient derivatives

We implement forward AD for det. We also simplify the implementation of
the backward, and leave a note on how to implement it properly for
singular matrices. We leave thad for future work.

Note (by looking at the OpInfo) that the current implementation passes
the same tests as the one before. We skip the forward-over-backward in
the singular case, as that one was not working in the gradgrad case
either.

[ghstack-poisoned]

lezcano committed:
Update on "Add forward AD for linalg.det and simplify its backward"


This PR is in preparation for implementing `logdet` and `slogdet` as
structured kernels + implementing them with more efficient derivatives

We implement forward AD for det. We also simplify the implementation of
the backward, and leave a note on how to implement it properly for
singular matrices. We leave thad for future work.

Note (by looking at the OpInfo) that the current implementation passes
the same tests as the one before. We skip the forward-over-backward in
the singular case, as that one was not working in the gradgrad case
either.

[ghstack-poisoned]

lezcano committed:
Update on "Add forward AD for linalg.det and simplify its backward"


This PR is in preparation for implementing `logdet` and `slogdet` as
structured kernels + implementing them with more efficient derivatives

We implement forward AD for det. We also simplify the implementation of
the backward, and leave a note on how to implement it properly for
singular matrices. We leave thad for future work.

Note (by looking at the OpInfo) that the current implementation passes
the same tests as the one before. We skip the forward-over-backward in
the singular case, as that one was not working in the gradgrad case
either.

[ghstack-poisoned]

lezcano committed:
Update on "Add forward AD for linalg.det and simplify its backward"


This PR is in preparation for implementing `logdet` and `slogdet` as
structured kernels + implementing them with more efficient derivatives

We implement forward AD for det. We also simplify the implementation of
the backward, and leave a note on how to implement it properly for
singular matrices. We leave thad for future work.

Note (by looking at the OpInfo) that the current implementation passes
the same tests as the one before. We skip the forward-over-backward in
the singular case, as that one was not working in the gradgrad case
either.

[ghstack-poisoned]

nikitaved commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -4026,134 +4026,43 @@ Tensor linalg_matrix_exp_differential(
+    return at::linalg_lu_solve(LU_, pivots, d, /*left=*/true, /*adjoint=*/true);
+  } else {
+    // TODO In this case, the trick above just does not cut it
+    // The proper way of doing this is doing `auto mask = det == 0.;` and then
+    // if any determinant is zero, use an SVD decomposition to compute the
+    // derivative in those inputs (not all inputs). The derivative may be then
```
But this is what is done in the current code. Why would we remove that now?

nikitaved commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -4026,134 +4026,43 @@ Tensor linalg_matrix_exp_differential(
+    return at::linalg_lu_solve(LU_, pivots, d, /*left=*/true, /*adjoint=*/true);
+  } else {
+    // TODO In this case, the trick above just does not cut it
+    // The proper way of doing this is doing `auto mask = det == 0.;` and then
+    // if any determinant is zero, use an SVD decomposition to compute the
+    // derivative in those inputs (not all inputs). The derivative may be then
```
But this is what is done in the current code. What is it being removed now?

Lezcano replied:
Not quite. The point here is that:
- If there is just one singular value that is zero, then the derivative is not zero
- If there is more than one singular value that is zero, the derivative is zero
- The SVD returns the singular values in decreasing order, so if the det is zero, we can just remove the smallest singular value and compute the derivative as (det(U) * det(V).conj() * s_1 * ... * s_{n-1})u^Tv

The last point is the relevant one to perform this optimisation.

nikitaved replied:
OK, makes sense.

Lezcano replied:
I realised it's not that easy. It's easy to compute the derivative, but the second derivative will not be corrct.

I am not implementing it in this PR as I need to move on to do other things. I've left a detailed comment on what's the issue.

Lezcano replied:
Fwiw, the second derivative of a vector with more than one zero is not correct in `torch.prod` afaik, so it's fine if it doesn't work for `torch.det` either.

Lezcano replied:
Sorry, forgot to push the changes. Now they're in.

nikitaved commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -4026,134 +4026,43 @@ Tensor linalg_matrix_exp_differential(
+    // computed explicitly by noting that the gradient of the derivative of the
+    // determinant is given by adj(A^H)* grad where adj(A) is the adjugate of A.
+    // Then, the adjugate of a singular matrix may be computed as per
+    // https://nhigham.com/2020/06/16/what-is-the-adjugate-of-a-matrix/
+    // adj(A) = conj(det(V)) * det(U) * (S[:-1].prod()) * U[-1] @ Vh[-1]
```
`S[1:]`?

nikitaved commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -4026,134 +4026,43 @@ Tensor linalg_matrix_exp_differential(
+    // computed explicitly by noting that the gradient of the derivative of the
+    // determinant is given by adj(A^H)* grad where adj(A) is the adjugate of A.
+    // Then, the adjugate of a singular matrix may be computed as per
+    // https://nhigham.com/2020/06/16/what-is-the-adjugate-of-a-matrix/
+    // adj(A) = conj(det(V)) * det(U) * (S[:-1].prod()) * U[-1] @ Vh[-1]
```
`S[1:]`?

Lezcano replied:
`S[:-1]`, as we want to remove the smallest singular value, that is, the last one.

nikitaved replied:
Right, sorry, forgot the order :)

lezcano committed:
Update on "Add forward AD for linalg.det and simplify its backward"


This PR is in preparation for implementing `logdet` and `slogdet` as
structured kernels + implementing them with more efficient derivatives

We implement forward AD for det. We also simplify the implementation of
the backward, and leave a note on how to implement it properly for
singular matrices. We leave thad for future work.

Note (by looking at the OpInfo) that the current implementation passes
the same tests as the one before. We skip the forward-over-backward in
the singular case, as that one was not working in the gradgrad case
either.

[ghstack-poisoned]

lezcano committed:
Update on "Add forward AD for linalg.det and simplify its backward"


This PR is in preparation for implementing `logdet` and `slogdet` as
structured kernels + implementing them with more efficient derivatives

We implement forward AD for det. We also simplify the implementation of
the backward, and leave a note on how to implement it properly for
singular matrices. We leave thad for future work.

Note (by looking at the OpInfo) that the current implementation passes
the same tests as the one before. We skip the forward-over-backward in
the singular case, as that one was not working in the gradgrad case
either.

[ghstack-poisoned]

lezcano committed:
Update on "Add forward AD for linalg.det and simplify its backward"


This PR is in preparation for implementing `logdet` and `slogdet` as
structured kernels + implementing them with more efficient derivatives

We implement forward AD for det. We also simplify the implementation of
the backward, and leave a note on how to implement it properly for
singular matrices. We leave thad for future work.

Note (by looking at the OpInfo) that the current implementation passes
the same tests as the one before. We skip the forward-over-backward in
the singular case, as that one was not working in the gradgrad case
either.

[ghstack-poisoned]

lezcano committed:
Update on "Add forward AD for linalg.det and simplify its backward"


This PR is in preparation for implementing `logdet` and `slogdet` as
structured kernels + implementing them with more efficient derivatives

We implement forward AD for det. We also simplify the implementation of
the backward, and leave a note on how to implement it properly for
singular matrices. We leave thad for future work.

Note (by looking at the OpInfo) that the current implementation passes
the same tests as the one before. We skip the forward-over-backward in
the singular case, as that one was not working in the gradgrad case
either.

[ghstack-poisoned]

lezcano committed:
Update on "Add forward AD for linalg.det and simplify its backward"


This PR is in preparation for implementing `logdet` and `slogdet` as
structured kernels + implementing them with more efficient derivatives

We implement forward AD for det. We also simplify the implementation of
the backward, and leave a note on how to implement it properly for
singular matrices. We leave thad for future work.

Note (by looking at the OpInfo) that the current implementation passes
the same tests as the one before. We skip the forward-over-backward in
the singular case, as that one was not working in the gradgrad case
either.

[ghstack-poisoned]

nikitaved commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -4050,134 +4050,55 @@ Tensor linalg_matrix_exp_differential(
+    // prod_safe_zeros_backward is not the second derivative of the product.
+    // It is not clear to me how to implement the second derivative of the
+    // product efficently. Note that this is also currently a problem when we
+    // compute higher derivatives of `x.prod()` and `x` has more than one zero.
+    return at::linalg_solve(A.mH(), d);
   }
```
Do you know which samples `gradgradcheck` fails on?

Lezcano replied:
I filled up a follow-up PR https://github.com/pytorch/pytorch/pull/80217 to unlock this one (as this one is already working and these modifications may introduce some weird errors in some platforms)

lezcano committed:
Update on "Add forward AD for linalg.det and simplify its backward"


This PR is in preparation for implementing `logdet` and `slogdet` as
structured kernels + implementing them with more efficient derivatives

We implement forward AD for det. We also simplify the implementation of
the backward, and leave a note on how to implement it properly for
singular matrices. We leave thad for future work.

Note (by looking at the OpInfo) that the current implementation passes
the same tests as the one before. We skip the forward-over-backward in
the singular case, as that one was not working in the gradgrad case
either.

[ghstack-poisoned]

nikitaved commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -4050,134 +4050,55 @@ Tensor linalg_matrix_exp_differential(
+    // auto D = prod_safe_zeros_backward(alpha.unsqueeze(-1), S, S.dim() - 1);
+    // return (U * D.unsqueeze(-2)).matmul(Vh);
+    //
+    // The issue with this code is that the derivative given by autograd of
+    // prod_safe_zeros_backward is not the second derivative of the product.
+    // It is not clear to me how to implement the second derivative of the
+    // product efficently. Note that this is also currently a problem when we
```
What is the issue with the `prod`?

Lezcano replied:
It's a non-issue, sorry. This comment is removed in https://github.com/pytorch/pytorch/pull/80217

nikitaved reviewed (approved):
Thank you, Mario. LGTM! Let's see the output, and that of the next one in the stack.

albanD reviewed (approved):
SGTM

Lezcano commented:
@pytorchbot merge -g

Lezcano commented:
The error, while related to the backward of conv, seems completely unrelated and it looks like a flaky test cc @janeyx99 @jbschlosser (note that the previous CI run was green and there were no changes in this PR between that run and this one).

I'm going YOLO and trying a merge anyway. Wish me the best.
